src/com/google/enterprise/adaptor/opentext/OpentextAdaptor.java

```diff
@@ -657,134 +657,180 @@ public void getDocIds(DocIdPusher pusher) throws InterruptedException {
     if (!this.markAllDocsAsPublic) {
       MemberService memberService =
           this.soapFactory.newMemberService(documentManagement);
-      Map<GroupPrincipal, List<Principal>> groupDefinitions =
-          getGroups(memberService);
-      if (this.publicAccessGroupEnabled) {
-        List<Principal> publicAccessGroup = getPublicAccessGroup(memberService);
-        if (publicAccessGroup.size() > 0) {
-          groupDefinitions.put(
-              new GroupPrincipal("[Public Access]", this.localNamespace),
-              publicAccessGroup);
+      DocIdPusher.FeedType feedType = DocIdPusher.FeedType.FULL;
+      SOAPFaultException groupsException = null;
+      Groups groups = new Groups(memberService, this);
+      try {
+        groups.addGroups();
+        if (this.publicAccessGroupEnabled) {
+          groups.addPublicAccessGroup("[Public Access]", this.localNamespace);
         }
+        if (!groups.isFull()) {
+          feedType = DocIdPusher.FeedType.INCREMENTAL;
+        }
+      } catch (SOAPFaultException e) {
+        groupsException = e;
+        feedType = DocIdPusher.FeedType.INCREMENTAL;
+        log.log(Level.WARNING, "Error while reading groups", e);
       }
+
+      // If some groups were found, push them; otherwise, if an
+      // exception was thrown, rethrow it as before.
+      Map<GroupPrincipal, List<Principal>> groupDefinitions =
+          groups.getGroupDefinitions();
       if (groupDefinitions.size() > 0) {
-        pusher.pushGroupDefinitions(groupDefinitions, true);
+        if (feedType == DocIdPusher.FeedType.INCREMENTAL) {
+          log.log(Level.FINE, "Sending partial groups results");
+        }
+        pusher.pushGroupDefinitions(groupDefinitions,
+            /* case sensitive */ true, feedType,
+            /* group source */ null, /* exception handler */ null);
+      } else if (groupsException != null) {
+        throw groupsException;
       }
     }
   }
 
   @VisibleForTesting
-  Map<GroupPrincipal, List<Principal>> getGroups(MemberService memberService) {
-    Map<GroupPrincipal, List<Principal>> groupDefinitions =
-        new HashMap<GroupPrincipal, List<Principal>>();
-    MemberSearchOptions options = new MemberSearchOptions();
-    options.setFilter(SearchFilter.GROUP);
-    options.setColumn(SearchColumn.NAME);
-    options.setMatching(SearchMatching.STARTSWITH);
-    options.setSearch(""); // Empty string matches everything.
-    options.setScope(SearchScope.SYSTEM);
-    options.setPageSize(50);
-    PageHandle handle = memberService.searchForMembers(options);
-    if (handle == null) {
-      log.log(Level.WARNING, "No groups found");
+  static class Groups {
+    private MemberService memberService;
+    private OpentextAdaptor adaptor;
+    private Map<GroupPrincipal, List<Principal>> groupDefinitions;
+    private boolean isFull = true;
```

Comment:
Please clarify which thing is full. Is it about being a full or incremental push? or is it about this datastructure being at capacity?  Please also modfify the name of the method isFull()